eclipse/src/saros/editor/EditorActivityDelayer.java

```diff
@@ -0,0 +1,89 @@
+package saros.editor;
+
+import java.util.Objects;
+import org.apache.log4j.Logger;
+import org.eclipse.swt.widgets.Display;
+import saros.activities.EditorActivity;
+import saros.activities.TextEditActivity;
+import saros.activities.TextSelectionActivity;
+import saros.activities.ViewportActivity;
+import saros.session.AbstractActivityProducer;
+import saros.session.ISarosSession;
+import saros.ui.util.SWTUtils;
+import saros.util.StackTrace;
+
+/**
+ * This class is responsible for delaying and discarding certain activities that are normally
+ * generated frequently.
+ */
+// TODO Viewport activities are also fire to frequently, worst case is selecting text with mouse
+// while scrolling up or down
+
+final class EditorActivityDelayer extends AbstractActivityProducer {
+
+  private static final int DELAY = 100; // ms
+
+  private static final Logger log = Logger.getLogger(EditorActivityDelayer.class);
+
+  private final Display display;
+
+  private TextSelectionActivity currentSelectionActivity = null;
+
+  EditorActivityDelayer() {
+    display = SWTUtils.getDisplay();
+    Objects.requireNonNull(display, "CTOR invoked from non SWT Thread");
+  }
+
+  void start(final ISarosSession session) {
+    session.addActivityProducer(this);
+  }
+
+  void stop(final ISarosSession session) {
+    session.removeActivityProducer(this);
+  }
+
+  void fireActivity(final ViewportActivity activity) {
+    checkSWTAccess();
+    flushCurrentSelectionActivity();
+    super.fireActivity(activity);
+  }
+
+  void fireActivity(final TextSelectionActivity activity) {
+    checkSWTAccess();
+
+    if (currentSelectionActivity == null) {
+      currentSelectionActivity = activity;
+      display.timerExec(DELAY, this::flushCurrentSelectionActivity);
+    } else if (currentSelectionActivity.getPath().equals(activity.getPath())) {
+      currentSelectionActivity = activity;
+    } else {
+      flushCurrentSelectionActivity();
+      currentSelectionActivity = activity;
+      display.timerExec(DELAY, this::flushCurrentSelectionActivity);
+    }
+  }
+
+  void fireActivity(final TextEditActivity activity) {
+    checkSWTAccess();
+    flushCurrentSelectionActivity();
+    super.fireActivity(activity);
+  }
+
+  void fireActivity(final EditorActivity activity) {
+    checkSWTAccess();
+    flushCurrentSelectionActivity();
+    super.fireActivity(activity);
+  }
+
+  private void flushCurrentSelectionActivity() {
+    if (currentSelectionActivity == null) return;
+
+    super.fireActivity(currentSelectionActivity);
+    currentSelectionActivity = null;
+  }
+
+  private void checkSWTAccess() {
```

Comment:
Why do you check the access? Are you only allowed to fire activities from an SWT thread? If so, this should be mentioned in the javadoc of the class and the `fireActivity(...)` mehtod(s).

Reply:
Welcome to Eclipse where 99% of the time every code is called from SWT. Of course I could make this thread safe but it annoys me every time because I do not know if it is 100% SWT or just 99% SWT all the time.

Reply:
I did a little testing and I think it is fine. Changed it so activities are dropped. If this is really called from non SWT then it is likely that either some Eclipse stuff or another 3rd party plugin broke the threading rules (which is not uncommon in Eclipse).